src/value.cpp

```diff
@@ -411,7 +411,7 @@ namespace Exiv2 {
     {
         value_ = buf;
         // ensure count>0 and nul terminated # https://github.com/Exiv2/exiv2/issues/1484
-        if (value_.size() == 0 || value_[value_.size()-1] != '\0') value_ += '\0';
+        if (value_.size() == 0 || (value_.size() > 0 && value_[value_.size()-1] != '\0')) value_ += '\0';
```

Comment:
I still don't get why this extra check is needed, isn't || a short-circuit operator?

Reply:
If it's empty _(size()==0)_ we want to add a nul byte.  So "" (length 0) becomes "\0" (length 1)

If it's not empty, we add a nul byte if the last byte isn't nul.  So if it's "fred" (length 4), it becomes "fred\0" (length 5).  If it's "fred\0" (length 5) we do nothing.

Reply:
@kmilos: I think your confusion is caused by Exiv2 using std::string in StringValueBase.  The Ascii object in the Exif/Tiff spec is a nul terminated string.

Reply:
I'm fully aware of the cause of this bug.

What I don't get is why your change was needed and why it makes any difference at all, when [it is NOOP logically speaking from what I know about the || operator](https://en.wikipedia.org/wiki/Short-circuit_evaluation#Definition).

Reply:
But anyway, it works, and we should move on.

Reply:
I'm pretty sure @kmilos is right, I don't see how this change could have had any impact at all.   
Both versions are functionally equivalent.       
So I don't see how this could have fixed a previously broken behavior.

Reply:
We're all puzzled by this.  However the fact is that two issues are resolved:
```bash
$ exiv2 -M'set Exif.Image.Copyright ""' foo.jpg | exiv2 -g Copyright foo.jpg'
$ exiv2 -M'set Exif.Image.Copyright' foo.jpg | exiv2 -g Copyright foo.jpg'
```
Both these commands now _correctly_ report:

```bash
Exif.Image.Copyright                         Ascii       1 
```
The prior behaviour was `Ascii  0`.

Reply:
The code:
```cpp
if (value_.size() == 0 || value_[value_.size()-1] != '\0') value_ += '\0';
```
Could be more verbosely written as:
```cpp
if (value_.size() == 0 ) {
    value_ += '\0';
} else if (value_[value_.size()-1] != '\0') {
    value_ += '\0';
}
```

Reply:
I also get this behavior with just 
```cpp
if (value_.size() == 0 || value_[value_.size()-1] != '\0') value_ += '\0';
```

the compiler literally translates the two versions into [identical assembly](https://www.godbolt.org/z/G8caPo)

Reply:
... which is exactly how the short-circuit || in the original patch works (or is supposed to work), hence the bewilderment ;)

Reply:
You're right, guys.  This code has pointless redundancy.  I've reinstated Milos's code from 26ba26e241e10bcac91fac8ce3f0b5a4dd7a45cd and ran some manual tests.  Everything seems good.  The additional tests I added to tests/bugfixes/github/test_issue_1489.py in ba6a577ab901cf32846dfdc65e5216417f3b5507 can _and should_ remain.

Clearly _**operator error**_ in this report: https://github.com/Exiv2/exiv2/pull/1486#issuecomment-796753668 

I'll submit a PR in a few minutes.

Reply:
Thanks for checking so thoroughly Robin, good to know I'm not going crazy ;)

Some incremental build gremlin? I guess we'll never know...

Reply:
Guys.  Something has gone horribly wrong with both my machines and I'm unable to submit the PR.  Could somebody submit this please:

```patch
    int AsciiValue::read(const std::string& buf)
    {
        value_ = buf;
        // ensure count>0 and nul terminated # https://github.com/Exiv2/exiv2/issues/1484
        if (value_.size() == 0 || value_[value_.size()-1] != '\0') value_ += '\0';
        return 0;
    }
```